### PR TITLE
Symbol table

### PR DESCRIPTION
**Pull request by thom776g**

### CODE REVIEW

File: GOAT-Compiler/SymbolTableBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                default:
+                    throw new TypeAccessException();
+            }
+        }
+
+        
+
+        public override void OutAVarDecl(AVarDecl node)
+        {
+            _symboltable.AddSymbol(node.GetId().Text, _processTypeOfNode(node));
```
Comment (Mati-AAU) [inline GOAT-Compiler/SymbolTableBuilder.cs:56]: should process node.GetTypes

File: GOAT-Compiler/SymbolTableBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                    _symboltable.CloseScope();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds a string with the name of the right type to typeTable.
+        /// </summary>
+        /// <param name="node">The given node being anlysed</param>
+        public override void OutAIntTypes(AIntTypes node)
```
Comment (Mati-AAU) [inline GOAT-Compiler/SymbolTableBuilder.cs:124]: for all these types nodes: should add to node not node.parent

File: SymbolTableTest/SymbolTableTest.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            symbolTable.OpenScope();
+            symbolTable.OpenScope();
+            symbolTable.AddSymbol("a", Types.Integer);
+            symbolTable.CloseScope();
+            Assert.Null(symbolTable.GetSymbol("a"));
+            symbolTable.CloseScope();
+        }
+
+        [Fact]
+        public void Chekcs_The_Right_Symbol_Gets_Returned_When_In_A_Scope()
```
Comment (Mati-AAU) [inline SymbolTableTest/SymbolTableTest.cs:101]: typo, checks not chekcs

File: SymbolTableTest/SymbolTableTest.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            symbolTable.OpenScope();
+            symbolTable.AddSymbol("a", Types.Boolean);
+            Assert.Equal(Types.Boolean, symbolTable.GetSymbol("a").type);
+            symbolTable.CloseScope();
+            Assert.Equal(Types.Integer, symbolTable.GetSymbol("a").type);
+            symbolTable.CloseScope();
+        }
+
+        [Fact]
+        public void Symbol_Build_test()
```
Comment (Mati-AAU) [inline SymbolTableTest/SymbolTableTest.cs:115]: maybe just a quick comment on what this test does

File: GOAT-Compiler/Symbol.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public enum Types
+    {
+        Integer,
+        FloatingPoint,
+        Vector,
+        Boolean,
+        Void
+    };
+
+    internal class Symbol
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply GOAT-Compiler/Symbol.cs:19]: later, that is

**Review by Mati-AAU — CHANGES_REQUESTED** [review]

File: GOAT-Compiler/SymbolTableBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+using GOATCode.node;
+using System;
+using System.Collections.Generic;
+
+namespace GOAT_Compiler
+{
+    internal class SymbolTableBuilder : DepthFirstAdapter
+    {
+        
+        internal ISymbolTable _symboltable;
```
Comment (Mati-AAU) [inline GOAT-Compiler/SymbolTableBuilder.cs:11]: this should be private

**Review by Mati-AAU — CHANGES_REQUESTED** [review]

File: GOAT-Compiler/RecSymbolTable.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        public RecSymbolTable() 
+        {
+            tables = new Table(null);
+            globalScope = new Table(tables);
+            tables.ChildrenTables.Add(globalScope);
+            currentScope = tables;
+        }
+
+        public void OpenScope()
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:26]: mangler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                else
+                {
+                    currentScope.ChildrenTables.Add(new Table(currentScope));
+                    currentScope = currentScope.ChildrenTables[visitcount];
+
+                }
+            }
+        }
+
+        public void CloseScope()
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:51]: Mangler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            currentScope = currentScope.ParentTable;
+
+            if (currentScope.ParentTable == null)
+            {
+                buildComplete = true;
+            }
+
+        }
+
+        public Symbol GetSymbol(string Name)
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:63]: Managler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            return currentScope.GetSymbol(Name);
+        }
+
+        public void AddSymbol(string Name, Types type)
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:68]: Managler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            currentScope.SetSymbol(Name, type);
+        }
+
+        public bool IsComplete()
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:73]: Managler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+        public Table(Table Parent)
+        {
+            Symbols = new Dictionary<string, Symbol>();
+            ChildrenTables = new List<Table>();
+            ParentTable = Parent;
+            VisitCounter = 0;
+        }
+
+        public Symbol GetSymbol(string Name)
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:96]: Managler doc

File: GOAT-Compiler/RecSymbolTable.cs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+    class Table
+    {
+        public Table ParentTable { get; private set; }
+        public List<Table> ChildrenTables { get; set; }
+        private readonly Dictionary<string, Symbol> Symbols;
+
+        public int VisitCounter { get; set; }
+
+        public Table(Table Parent)
```
Comment (S3JER) [inline GOAT-Compiler/RecSymbolTable.cs:88]: Managler doc

File: GOAT-Compiler/Assembly.cs [hunk 14]
```diff
@@ -0,0 +1,4 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("SymbolTableTest")]
+[assembly: InternalsVisibleTo("PrettyPrintATestFile")]
```
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/Assembly.cs:4]: You are quite right

File: GOAT-Compiler/ISymbolTable.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        /// Adds a symbol to the symbol table.
+        /// If the same symbol exists in an outer scope it is temporarily removed from the table and stored in the new symbol.
+        /// In case of a duplicate symbol an exception is thrown.
+        /// </summary>
+        /// <param name="Name">The name of the symbol</param>
+        /// <param name="Type">The type of the symbol</param>
+        /// <exception cref="ArgumentException">Exception that is thrown if there is a duplicate definition.</exception>
+        public void AddSymbol(string Name, Types Type);
+
+        public bool IsComplete();
```
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/ISymbolTable.cs:36]: Done

File: GOAT-Compiler/RecSymbolTable.cs [hunk 16]
```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GOAT_Compiler
+{
+    class RecSymbolTable : ISymbolTable
```
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/RecSymbolTable.cs:9]: idk

File: GOAT-Compiler/Symbol.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/Symbol.cs:19]: yes

File: GOAT-Compiler/RecSymbolTable.cs (hunk 12 above)
Comment (Casper-NS) [reply GOAT-Compiler/RecSymbolTable.cs:109]: done

File: GOAT-Compiler/RecSymbolTable.cs (hunk 13 above)
Comment (Casper-NS) [reply GOAT-Compiler/RecSymbolTable.cs:96]: done

**Review by Mati-AAU — APPROVED** [review]

**Pull request merged**